libexec/scoop-shim.ps1

```diff
@@ -0,0 +1,227 @@
+. "$PSScriptRoot\..\lib\getopt.ps1" # for getopt
+. "$PSScriptRoot\..\lib\install.ps1" # for rm_shim
+
+$opt, $addArgs, $err = getopt $Args 'g' 'global'
+if ($err) { abort "scoop shim: $err" 1 }
+$global = $opt.g -or $opt.global
+$globalTag = if ($global) { 'global' } else { 'local' }
+
+$shimName = $addArgs[0]
+
+if ($SubCommand -notin @('add', 'remove', 'rm', 'list', 'info', 'alter')) {
+    'ERROR: <subcommand> must be one of: add, remove/rm, list, info, alter'
+    my_usage
+    exit 1
+}
+
+if (!$shimName -and $SubCommand -ne 'list') {
+    "ERROR: <shim_name> must be specified for subcommand '$SubCommand'"
+    my_usage
+    exit 1
+}
+
+if ($SubCommand -eq 'add' -and $null -eq $addArgs[1]) {
+    "ERROR: <command_path> must be specified for subcommand 'add'"
+    my_usage
+    exit 1
+}
+
+if (-not (Get-FormatData ScoopShims)) {
+    Update-FormatData "$PSScriptRoot\..\supporting\formats\ScoopTypes.Format.ps1xml"
+}
+
+$localShimDir = shimdir $false
+$globalShimDir = shimdir $true
+
+function Get-ShimInfo($ShimPath) {
+    $info = [Ordered]@{}
+    $info.Name = strip_ext (fname $ShimPath)
+    $info.Path = $ShimPath -replace 'shim$', 'exe'
+    $info.Source = (get_app_name_from_shim $ShimPath) -replace '^$', 'External'
+    $info.Type = if ($ShimPath.EndsWith('.ps1')) { 'ExternalScript' } else { 'Application' }
+    $altShims = Get-Item -Path "$ShimPath.*" -Exclude '*.shim', '*.cmd', '*.ps1'
+    if ($altShims) {
+        $info.Alternatives = (@($info.Source) + ($altShims | ForEach-Object { $_.Extension.Remove(0, 1) } | Select-Object -Unique)) -join ' '
+    }
+    $info.IsGlobal = $ShimPath.StartsWith("$globalShimDir")
+    $info.IsHidden = !((Get-Command -Name $info.Name).Path -eq $info.Path)
+    [PSCustomObject]$info
+}
+
+function Get-ShimPath($shimName, $Global) {
+    '.shim', '.ps1' | ForEach-Object {
+        $shimPath = Join-Path (shimdir $Global) "$shimName$_"
+        if (Test-Path $shimPath) {
+            return $shimPath
+        }
+    }
+}
+
+function Get-ShimTarget($ShimPath) {
+    if ($ShimPath) {
+        $shimTarget = if ($ShimPath.EndsWith('.shim')) {
+            (Get-Content -Path $ShimPath | Select-Object -First 1).Replace('path = ', '')
+        } else {
+            ((Select-String -Path $ShimPath -Pattern '^(?:@rem|#)\s*(.*)$').Matches.Groups | Select-Object -Index 1).Value
+        }
+        if (!$shimTarget) {
+            $shimTarget = ((Select-String -Path $ShimPath -Pattern '[''"]([^@&]*?)[''"]' -AllMatches).Matches.Groups | Select-Object -Last 1).Value
+        }
+        $shimTarget | Convert-Path
+    }
+}
+
+switch ($SubCommand) {
+    'add' {
+        $commandPath = $addArgs[1]
+        if ($commandPath -match "^[`"'](.*?)[`"']\s*(.*)$") {
+            $commandPath = $Matches[1]
+            $commandArgs = $Matches[2]
+        } elseif ($commandPath -match '^([^\s]*)\s*(.*)$') {
+            $commandPath = $Matches[1]
+            $commandArgs = $Matches[2]
+        }
```

Comment:
I'll try to handle `-g` by the script itself.

Reply:
Yes that'll be much better

Reply:
It is done, but still has bug in CMD. I'll try to fix it in this weekend.

Reply:
Yes it does, but what if there's another `-g/--global` switch in your shimmed program? I forced `-g` to be latest one, otherwise it will be used as shimmed command line.

Reply:
1. If the shimmed program needs `--global`, it can be passed as is because PS does not recognize flags with 2 hyphens.
2. If the shimmed program needs `-g` or `-global`, just put those in quotes. Then they will be passed to the shim.

See this:

![image](https://user-images.githubusercontent.com/46838874/154889604-57420944-629e-419f-9881-3f8abcaaf978.png)

![image](https://user-images.githubusercontent.com/46838874/154889621-4369be82-9c2d-486b-b6e4-7ff06fe656d1.png)


The value of $global is false in all cases, which is what we need.

Reply:
Use a direct approach to handle `-g/--global` flag. In syntax, explicitly use global flag in last position, and the script check `$Args[-1]` to see if manipulate global shim.

Test passed in PowerShell and CMD, but for bash, whitespaces in path could not be figured out while `$Args` are separated by whitespaces and no quote is added in `$Args`'s items.

Reply:
Do CMD and Bash behave the same with my method (in comment above) also?

Also could you also confirm for Git Bash (it is slightly different than WSL Bash)?

Reply:
It's not related to the arguments parsing, it's COMMAND_PATH parsing... If there are whitespaces in command path, WSL got error.

For arguments, both methods work, and is equivalent IMO.

Reply:
Tested in MSYS2 and Cygwin, both got errors as in WSL.

Reply:
> Yes it does, but what if there's another `-g/--global` switch in your shimmed program? I forced `-g` to be latest one, otherwise it will be used as shimmed command line. And if you use `-g` in command path (as arg) and want to make the shim global, `-g -g` is needed.

I did not fully understand this part. Could you give examples of the different cases?

Reply:
e.g., assume `test` has a `-g` flag, and if you want to shim `test -g` as global one, you could use `scoop shim add my-test test -g -g` (first `-g` is for `test`, latest one is for scoop shim).

A more complicated case: if `test`'s `-g` flag could accept argument (`test -g arg`), if you use `scoop shim add my-test test -g arg` to add local shim, the `-g` could not be marked as scoop shim's global flag, and you must use `scoop shim add my-test test -g arg -g` to add global one.

So I force `scoop shim`'s global flag (`-g`/`--global`) to be latest one to differ it from command path's `-g` (if existed).

Reply:
> But this could lead to an unexpected case that the only param in command is `-g` and you want a local shim, since `scoop shim add my-test test -g` will make a global shim, and in such situation, scoop could not distinguish command param with scoop shim param... This is a rare use case, IMO.

Yes, this was exactly what I was confused about. Even if it is rare, current implementation will always create a global shim irrespectively.

That is why I suggest using the method I described above. It has the following advantages:
- it is unambiguous. If shim needs `-g`, just pass it in quotes, like `'g'`.
- it does not put restriction that the global flag needs to be the last argument.

Reply:
The param syntax is not consistent with other scoop command if using `[Switch]`, just as I commented under `scoop info` patch.

A trick to set `test -g` as local shim is `scoop shim add my-shim test -g ''`, i.e., pass an empty param as latest param.

Reply:
> The param syntax is not consistent with other scoop command if using `[Switch]`, just as I commented under `scoop info` patch.

This is just an implementation detail; users do not need to know how params are implemented internally.

> A trick to set `test -g` as local shim is `scoop shim add my-shim test -g ''`, i.e., pass an empty param as latest param.

I'm personally against tricks/hacks like these, and this behaviour is not user friendly. `scoop shim add my-shim test -g` for global shim, and `scoop shim add my-shim test '-g'` for a local shim with g flag is much more intuitive for a user.

We are implementing this functionality from scratch, so why not do it in a proper way instead of adding workarounds?

Reply:
Personally, my only consideration is param syntax. If we use PS switch `-Global`, the `-g` is just a short form of it, and PS thinks `-g`, `-gl`, `-glo`, etc. all as `-global`, should this cause some problems? And PS doesn't support `--global` which needs a special treatment.

Reply:
> Personally, my only consideration is param syntax. If we use PS switch `-Global`, the `-g` is just a short form of it, and PS thinks `-g`, `-gl`, `-glo`, etc. all as `-global`, should this cause some problems? 

All apps other than PS use either `-g` or `--g[full name]`, so I don't think `-gl` or `-glo` should be a problem.

> And PS doesn't support `--global` which needs a special treatment.

Hmm, yes. But we can just mention:
```
#     -g, -global           Create a global shim
```

instead of 
```
#     -g, --global          Create a global shim
```

in the help text, which the users can see and use.

Reply:
I'll try to find a way to handle such param, or maybe mix the two methods.

Reply:
Some minor tweaks, and result is fine:

![image](https://user-images.githubusercontent.com/5832170/155119052-ea768a7b-a6c8-4244-95a9-783d7463cd07.png)

Reply:
Yes, I only added `[switch]$g` to avoid `-g*` mismatch, and handle `--global` in the script.